runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/BoundedTrieData.java

```diff
@@ -0,0 +1,585 @@
+    }
+    if (this.root != null) {
+      builder.setRoot(this.root.toProto());
+    }
+    return builder.build();
+  }
+
+  /** Creates a {@link BoundedTrieData} instance from its proto {@link BoundedTrie}. */
+  public static BoundedTrieData fromProto(BoundedTrie proto) {
+    if (proto.hasRoot()) {
+      return new BoundedTrieData(
+          null, BoundedTrieNode.fromProto(proto.getRoot()), proto.getBound());
+    } else {
+      return new BoundedTrieData(proto.getSingletonList(), null, proto.getBound());
+    }
+  }
+
+  /** Returns this {@link BoundedTrieData} as a {@link BoundedTrieNode}. */
+  @Nonnull
+  private synchronized BoundedTrieNode asTrie() {
+    if (this.root != null) {
+      return this.root;
+    } else {
+      BoundedTrieNode trieNode = new BoundedTrieNode();
+      if (this.singleton != null) {
+        trieNode.add(this.singleton);
+      }
+      return trieNode;
+    }
+  }
+
+  /** Returns a new {@link BoundedTrieData} instance that is a deep copy of this instance. */
+  public synchronized BoundedTrieData getCumulative() {
+    List<String> singleton = this.singleton == null ? null : new ArrayList<>(this.singleton);
+    // deep copy
+    BoundedTrieNode root = this.root == null ? null : this.root.deepCopy();
+    return new BoundedTrieData(singleton, root, this.bound);
+  }
+
+  /**
+   * Returns an immutable set of lists, where each list represents a path in the bounded trie. The
+   * last element in each path is a boolean in string representation denoting whether this path was
+   * truncated. i.e. <["a", "b", "false"], ["c", "true"]>
+   *
+   * @return The set of paths.
+   */
+  public synchronized Set<List<String>> getResult() {
+    if (this.root == null) {
+      if (this.singleton == null) {
+        return ImmutableSet.of();
+      } else {
+        List<String> list = new ArrayList<>(this.singleton);
+        list.add(String.valueOf(false));
+        return ImmutableSet.of(list);
+      }
+    } else {
+      return ImmutableSet.copyOf(this.root.flattened());
+    }
+  }
+
+  /**
+   * Adds a new path to this {@link BoundedTrieData} and hence the {@link BoundedTrieData} is
+   * modified.
+   *
+   * @param segments The path to add.
+   */
+  public synchronized void add(Iterable<String> segments) {
+    List<String> segmentsParts = ImmutableList.copyOf(segments);
+    if (this.root == null) {
+      if (this.singleton == null || !this.singleton.equals(segmentsParts)) {
+        this.root = this.asTrie();
+        this.singleton = null;
+      }
+    }
+
+    if (this.root != null) {
+      this.root.add(segmentsParts);
+      if (this.root.getSize() > this.bound) {
+        this.root.trim();
+      }
+    }
+  }
+
+  /**
+   * Combines this {@link BoundedTrieData} with another {@link BoundedTrieData} by doing a deep
+   * copy.
+   *
+   * @param other The other {@link BoundedTrieData} to combine with.
+   */
+  public synchronized void combine(@Nonnull BoundedTrieData other) {
```

Comment:
This return a void rather than a reference to updated object because the object is mutable and returning a reference from here will then require unnecessary overhead of using AtomicReference in BoundedTrieCell to safely be able to handle combine call from mutli-threads.